### PR TITLE
Add CornerNet benchmark

### PR DESCRIPTION
**Pull request by tianyuandu**
Also fix some bugs in training.

### CODE REVIEW

File: configs/cornernet/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+| Backbone        | Batch Size | Step/Total Epochs | Mem (GB) | Inf time (fps) | box AP | Pytorch Version | Download |
+| :-------------: | :--------: |:----------------: | :------: | :------------: | :----: | :-------------: | :------: |
+| HourglassNet-104 | 5 x 10 | 180/210 | 13.9 | | 41.3 | 1.5 | |
+| HourglassNet-104 | 6 x 8 | 180/210 | 15.9 | | 41.2 | 1.5 | |
+| HourglassNet-104 | 6 x 16 | 180/210 | 15.9 | | 41.1 | 1.5 | |
+| HourglassNet-104 | 3 x 32 | 180/210 | 9.5 | | 40.4 | 1.3 | |
+
+Note: 
+- All models are tested with TTA (single-scale and `flip=True`).
```
Comment (hellock) [inline configs/cornernet/README.md:25]: To have a fair comparison with other models, we need also report single-scale testing results.

File: configs/cornernet/cornernet_hourglass104_aug_test_batch3.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
+_base_ = ['../_base_/default_runtime.py']
```
Comment (hellock) [inline configs/cornernet/cornernet_hourglass104_aug_test_batch3.py:1]: The config filenames should follow the [naming convention](https://mmdetection.readthedocs.io/en/latest/config.html#config-name-style).

File: mmdet/models/detectors/cornernet.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                 neck,
+                 bbox_head,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        super(CornerNet, self).__init__(backbone, neck, bbox_head, train_cfg,
+                                        test_cfg, pretrained)
+
+    def merge_aug_results(self, aug_results, img_metas):
+        recovered_bboxes, aug_labels = [], []
```
Comment (ZwwWayne) [inline mmdet/models/detectors/cornernet.py:22]: Docstrings

File: mmdet/models/detectors/cornernet.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import torch
+
+from mmdet.core import bbox2result, bbox_mapping_back
+from ..builder import DETECTORS
+from .single_stage import SingleStageDetector
+
+
+@DETECTORS.register_module()
+class CornerNet(SingleStageDetector):
+
```
Comment (ZwwWayne) [inline mmdet/models/detectors/cornernet.py:10]: Docstrings

File: mmdet/models/detectors/cornernet.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        bboxes = torch.cat(recovered_bboxes, dim=0)
+        labels = torch.cat(aug_labels)
+
+        out_bboxes, out_labels = self.bbox_head._bboxes_nms(
+            bboxes, labels, self.bbox_head.test_cfg)
+
+        return out_bboxes, out_labels
+
+    def aug_test(self, imgs, img_metas, rescale=False):
+        img_inds = list(range(len(imgs)))
```
Comment (ZwwWayne) [inline mmdet/models/detectors/cornernet.py:42]: Docstrings

File: mmdet/models/detectors/cornernet.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import torch
+
+from mmdet.core import bbox2result, bbox_mapping_back
+from ..builder import DETECTORS
+from .single_stage import SingleStageDetector
+
+
+@DETECTORS.register_module()
+class CornerNet(SingleStageDetector):
```
Comment (ZwwWayne) [inline mmdet/models/detectors/cornernet.py:9]: We need a unit test to test the inference and training of this detector.

File: configs/cornernet/cornernet_hourglass104_8x6_210e_coco.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+_base_ = ['../_base_/default_runtime.py']
```
[the remark this answers is not in the document]
Comment (tianyuandu) [reply configs/cornernet/cornernet_hourglass104_8x6_210e_coco.py:1]: I think it's not necessary. `dataset_type`, `data_root` must be declared for defining `data`. `img_norm_cfg` must be declared for defining `train_pipeline` and `test_pipeline`. Also our `data` has different `samples_per_gpu`, `workers_per_gpu` and pipelines. At the end we only keep `evaluation` unchanged and using `_base_/coco_detection.py` could only omit the `evaluation` key.

[the remark this answers is not in the document]
Comment (ZwwWayne) [reply configs/cornernet/cornernet_hourglass104_8x6_210e_coco.py:1]: The following setting is already defined in `coco_detection.py` and do not need to  write again. You only need to specify `pipeline`, `samples_per_gpu`, `workers_per_gpu`.
```python  
type=dataset_type
ann_file=data_root + 'annotations/instances_train2017.json'
img_prefix=data_root + 'train2017/'
```

File: configs/cornernet/cornernet_hourglass104_8x6_210e_coco.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+test_cfg = dict(
+    corner_topk=100,
+    local_maximum_kernel=3,
+    distance_threshold=0.5,
+    score_thr=0.05,
+    max_per_img=100,
+    nms_cfg=dict(type='soft_nms', iou_threshold=0.5, method='gaussian'))
+# optimizer
+optimizer = dict(type='Adam', lr=0.0005)
+optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
```
Comment (ZwwWayne) [inline configs/cornernet/cornernet_hourglass104_8x6_210e_coco.py:109]: Is `grad_clip` necessary for CornerNet? If not, remove it.

File: configs/cornernet/README.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+}
+```
+
+## Results and models
+
+| Backbone        | Batch Size<br>gpus x images\_per\_gpu | Step/Total Epochs | Mem (GB) | Inf time (fps) | box AP | Pytorch Version | Download |
+| :-------------: | :--------: |:----------------: | :------: | :------------: | :----: | :-------------: | :------: |
+| HourglassNet-104 | [10 x 5](./cornernet_hourglass104_mstest_10x5_210e_coco.py)<br>align paper, need 2 V100 nodes | 180/210 | 13.9 | | 40.6<br>TTA: 41.3 | 1.5 | |
+| HourglassNet-104 | [8 x 6](./cornernet_hourglass104_mstest_8x6_210e_coco.py)<br>on single V100 node| 180/210 | 15.9 | | 40.9<br>TTA: 41.2 | 1.5 | |
+| HourglassNet-104 | [32 x 3](./cornernet_hourglass104_mstest_32x3_210e_coco.py)<br>for 1080TI, need 4 nodes | 180/210 | 9.5 | | 39.6<br>TTA: 40.4 | 1.3 | |
```
Comment (hellock) [inline configs/cornernet/README.md:21]: The table looks quite complex. We can leave some settings described below the table. PyTorch version can be omitted.

File: configs/cornernet/cornernet_hourglass104_8x6_210e_coco.py (hunk 8 above)
Comment (tianyuandu) [reply configs/cornernet/cornernet_hourglass104_8x6_210e_coco.py:109]: I didn't implement any experiment without `grad_clip`, so I think it's better to keep it.

**Comment by hellock** [comment]
LGTM. The inference time is needed and then @xvjiarui may help upload the checkpoints.

File: configs/cornernet/README.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+| :-------------: | :--------: |:----------------: | :------: | :------------: | :----: | :------: |
+| HourglassNet-104 | [10 x 5](./cornernet_hourglass104_mstest_10x5_210e_coco.py) | 180/210 | 13.9 | 4.2 | 40.6<br>TTA: 41.3 | |
+| HourglassNet-104 | [8 x 6](./cornernet_hourglass104_mstest_8x6_210e_coco.py) | 180/210 | 15.9 | 4.2 | 40.9<br>TTA: 41.2 | |
+| HourglassNet-104 | [32 x 3](./cornernet_hourglass104_mstest_32x3_210e_coco.py) | 180/210 | 9.5 | 3.9 | 39.6<br>TTA: 40.4 | |
+
+Note: 
+- TTA setting is single-scale and `flip=True`.
+- Experiments with `images_per_gpu=6` are implemented on Tesla V100-SXM2-32GB, `images_per_gpu=3` are implemented on GeForce GTX 1080 Ti.
+- Here are the descriptions of each experiment setting:
+    - 10 x 5: 10 GPUs with 5 images per gpu. This is the setting align to paper and needs 2 nodes to train.
```
Comment (ZwwWayne) [inline configs/cornernet/README.md:27]: This is the setting align to paper and needs 2 nodes to train. -> This is the same setting as that reported in the original paper.

File: configs/cornernet/README.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+| Backbone        | Batch Size | Step/Total Epochs | Mem (GB) | Inf time (fps) | box AP | Download |
+| :-------------: | :--------: |:----------------: | :------: | :------------: | :----: | :------: |
+| HourglassNet-104 | [10 x 5](./cornernet_hourglass104_mstest_10x5_210e_coco.py) | 180/210 | 13.9 | 4.2 | 40.6<br>TTA: 41.3 | |
+| HourglassNet-104 | [8 x 6](./cornernet_hourglass104_mstest_8x6_210e_coco.py) | 180/210 | 15.9 | 4.2 | 40.9<br>TTA: 41.2 | |
+| HourglassNet-104 | [32 x 3](./cornernet_hourglass104_mstest_32x3_210e_coco.py) | 180/210 | 9.5 | 3.9 | 39.6<br>TTA: 40.4 | |
+
+Note: 
+- TTA setting is single-scale and `flip=True`.
+- Experiments with `images_per_gpu=6` are implemented on Tesla V100-SXM2-32GB, `images_per_gpu=3` are implemented on GeForce GTX 1080 Ti.
```
Comment (ZwwWayne) [inline configs/cornernet/README.md:25]: Experiments with `images_per_gpu=6` are conducted on Tesla V100-SXM2-32GB, `images_per_gpu=3` are conducted on GeForce GTX 1080 Ti.

File: mmdet/models/detectors/cornernet.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    """CornerNet: Detecting Objects as Paired Keypoints.
```
Comment (ZwwWayne) [inline mmdet/models/detectors/cornernet.py:10]:
```python
    """CornerNet.

    This detector is the implementation of the paper `CornerNet: Detecting Objects as Paired Keypoints
    <https://arxiv.org/abs/1808.01244>`_ .
    """

```

File: mmdet/models/detectors/cornernet.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        Note:
+            ``imgs`` must including flipped image pairs.
+
+        Returns:
+            bbox_results (tuple[np.ndarry]): Detection result of each class.
+        """
+        img_inds = list(range(len(imgs)))
+
+        assert img_metas[0][0]['flip'] + img_metas[1][0]['flip'], (
+            "aug test must have flipped image pair")
```
Comment (ZwwWayne) [inline mmdet/models/detectors/cornernet.py:79]: Use single quote rather than double quote when double quote is unnecessary.

**Pull request merged**